### PR TITLE
PAL-939 streaming client (archived, see V2 of this pr)

### PR DESCRIPTION
**Pull request by dbz120896**
Please review the V2 of this PR: https://github.com/gchq/Palisade-clients/pull/53

### CODE REVIEW

File: client-java/src/main/java/uk/gov/gchq/palisade/client/Client.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2020-2021 Crown Copyright
```
Comment (developer01189998819991197253) [inline client-java/src/main/java/uk/gov/gchq/palisade/client/Client.java:2]: Replace copyright template with Copyright 2018-2021 Crown Copyright for all Palisade files

**Review by developer01189998819991197253 — CHANGES_REQUESTED** [review]
first pass

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/WsEndpointFilteredResource.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    /**
+     * Called when the websocket closes
+     *
+     * @param session The websocket session that is to close
+     */
+    @OnClose
+    public void onClose(final WebSocketSession session) {
+        LOGGER.debug("<-- CLOSE request received", session.getId());
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/WsEndpointFilteredResource.java:164]: done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/MessageType.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    /**
+     * The return message is an error
+     */
+    ERROR,
+
+    /**
+     * The return message is a resource
+     */
+    RESOURCE;
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/MessageType.java:33]: I wish Oracle would just allow one way or the other and make the other one an error.
Done.

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/dft/DefaultQueryResponse.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                    createMessage(wsm).ifPresent(emitter::onNext);
+
+                }
+
+                if (emitter.isCancelled()) {
+                    LOGGER.debug("emitter.cancelled");
+                    return;
+                }
+
+            } while (!leave);
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/dft/DefaultQueryResponse.java:107]: I created the `createMessage` function to reduce complexity as there were too many exits.The check for a `CompleteMessage` should set it to true and then perform an else.
Done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/DefaultClientTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+class DefaultClientTest {
+
+    @BeforeEach
+    void setUp() throws Exception {
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+    }
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/DefaultClientTest.java:37]: Deleted.

File: client-java/src/main/java/uk/gov/gchq/palisade/client/util/Util.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+     */
+    public static Map<String, Object> substituteVariables(final Map<String, Object> input) {
+        var result = new HashMap<>(input);
+        for (Entry<String, Object> entry : input.entrySet()) {
+            var expr = entry.getValue().toString().trim();
+            if (expr.startsWith("${") && expr.endsWith("}")) {
+                var key = expr.substring(2, expr.length() - 1);
+                var repl = input.get(key);
+                if (repl != null) {
+                    result.put(entry.getKey(), repl);
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/util/Util.java:104]: done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/resource/WebSocketMessage.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+package uk.gov.gchq.palisade.client.internal.resource;
+
+import uk.gov.gchq.palisade.client.util.ImmutableStyle;
+
+import java.io.Serializable;
+import java.util.Map;
+import java.util.function.UnaryOperator;
+
+/**
+ * A websocket message is the actuial instance sent/received to/from the
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/resource/WebSocketMessage.java:25]: done

File: NOTICES.md [hunk 8]
```diff
@@ -1,17 +1,41 @@
...
+* AssertJ fluent assertions ([org.assertj:assertj-core:3.17.2](http://assertj.org/assertj-core))
+* Awaitility ([org.awaitility:awaitility:4.0.3](http://awaitility.org))
+* EventBus ([org.greenrobot:eventbus:3.2.0](http://greenrobot.org/eventbus/))
+* org.immutables.value ([org.immutables:value:2.8.2](http://immutables.org/value))
 * Spring Cloud Starter OpenFeign ([org.springframework.cloud:spring-cloud-starter-openfeign:2.2.0.RELEASE](https://projects.spring.io/spring-cloud))
 * clients-common ([uk.gov.gchq.palisade:clients-common:0.5.0-SNAPSHOT](https://github.com/gchq/Palisade-clients/tree/develop/clients-common))
-* common ([uk.gov.gchq.palisade:common:0.5.0-SNAPSHOT](https://github.com/gchq/Palisade-common))
-* readers-common ([uk.gov.gchq.palisade:readers-common:0.5.0-SNAPSHOT](https://github.com/gchq/Palisade-readers/tree/develop/readers-common))
+* common ([uk.gov.gchq.palisade:common:0.5.0-20210121.151104-22](https://github.com/gchq/Palisade-common))
+* readers-common ([uk.gov.gchq.palisade:readers-common:0.5.0-20210126.160626-27](https://github.com/gchq/Palisade-readers/tree/develop/readers-common))
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply NOTICES.md:26]: Yup, will do.

File: NOTICES.md [hunk 9]
```diff
@@ -1,17 +1,41 @@
...
+### [Eclipse Public License v2.0](./licenses/not recognised):
+* JUnit Jupiter (Aggregator) ([org.junit.jupiter:junit-jupiter:5.7.0](https://junit.org/junit5/))
+
+### [GNU Lesser General Public License 2.1](./licenses/gnu_lgpl_2.1.html):
+* Logback Classic Module ([ch.qos.logback:logback-classic:1.2.3](http://logback.qos.ch/logback-classic))
 
 ### [MIT License](./licenses/mit_license.txt):
 * Mockito ([org.mockito:mockito-all:1.10.19](http://www.mockito.org))
 * SLF4J API Module ([org.slf4j:slf4j-api:1.7.25](http://www.slf4j.org))
+* SLF4J API Module ([org.slf4j:slf4j-api:1.7.26](http://www.slf4j.org))
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply NOTICES.md:41]: client-java uses 1.7.26. I think other projects in the reactor use previous version. The client-java project has nothing to do with other client projects, so should be split into it's own repository.

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/internal/download/DownloaderTest.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @BeforeEach
+    void setup() {
+        this.url = BASE_URL + embeddedServer.getPort();
+        this.downloader = new Downloader(objectMapper, ENDPOINT);
+    }
+
+    @Test
+    void testSuccessfulDownload() throws Exception {
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/internal/download/DownloaderTest.java:62]: done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/internal/resource/ResourceClientTest.java [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        this.resourceClient = WebSocketClient
+            .createResourceClient(b -> b
+                .token(TOKEN)
+                .baseUri("ws://localhost:" + port + "/cluster/filteredResource/name/%t")
+                .objectMapper(objectMapper))
+            .connect();
+    }
+
+    @Test
+    void testMessageFlow() throws Exception {
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/internal/resource/ResourceClientTest.java:64]: done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/ClientManager.java [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+     *
+     * @param url a palisade url of the form pal:subprotocol:subname
+     * @return a session for the provided {@code url}
+     * @throws ClientException if a Palisade access error occurs or the url is
+     *                         invalid
+     */
+    public static Session openSession(final String url) {
+        var client = getClient(url);
+        return client.connect(url, Map.of());
+    }
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/ClientManager.java:81]: When running in a prod environment, you will not need to supply a properties object. User/port will be supplied on the url. I've added a test that exercises it.

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/WsEndpointFilteredResource.java [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+    public void onOpen(final String token, final WebSocketSession session) {
+        session.put(TOKEN_KEY, token);
+        this.messages = new ResourceGenerator(token, embeddedServer.getPort()).iterator();
+        if (!messages.hasNext()) {
+            sendComplete(session);
+        }
+        LOGGER.debug("WebSocket Server opened");
+    }
+
+    boolean completeSent = false;
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/WsEndpointFilteredResource.java:134]: That was a Klingon ;) Removed.

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+        var publisher = query
+            .execute()
+            .thenApply(QueryResponse::stream)
+            .get();
+
+        Flowable.fromPublisher(FlowAdapters.toPublisher(publisher))
+            .filter(m -> m.getType() == MessageType.RESOURCE)
+            .map(Resource.class::cast)
+            .map(session::fetch)
+            .subscribe(new FlowableSubscriber<Download>() {
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java:119]: Done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+                public void onNext(final Download t) {
+                    LOGGER.debug("## Got message: {}", t);
+                    var is = t.getInputStream();
+                    try {
+                        LOGGER.debug("## reading bytes");
+                        var ba = is.readAllBytes();
+                        LOGGER.debug("## read {} bytes", ba.length);
+                    } catch (Throwable e) {
+                        LOGGER.error("Got error reading inputstream into byte array", e);
+                        throw new IllegalStateException("Got error reading inputstream into byte array", e);
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java:131]: done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /*
+     * This tests that the download inputstream can be consumed within the same
+     * Flowable. But, there is a problem with Java 11:
+     * https://bugs.openjdk.java.net/browse/JDK-8228970 You can view the test here :
+     * https://hg.openjdk.java.net/jdk/jdk/rev/e4cc5231ce2d
+     */
+    @Test
+    @Disabled
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java:100]: I'll create one...... Done : https://jira.tdx.oneit.gov.uk/browse/PAL-979

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/Configuration.java [hunk 17]
```diff
@@ -0,0 +1,411 @@
...
+        return Optional.ofNullable(getProperty(properties, key, clazz));
+    }
+
+    private static Map<String, Object> process(final Map<String, Object> properties) {
+
+        // process the uri
+
+        String url = getProperty(properties, KEY_SERVICE_URL);
+
+        // if the url is not set, then we cannot propcess it
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/Configuration.java:308]: done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/WsEndpointFilteredResource.java [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+     * @param session The websocket session that is to close
+     */
+    @OnClose
+    public void onClose(final WebSocketSession session) {
+        LOGGER.debug("<-- CLOSE request received", session.getId());
+    }
+
+    private static final void sendComplete(final WebSocketSession session) {
+        send(session, Item.createMessage(b -> b
+            .putHeader("token", session.get(TOKEN_KEY, String.class).get())
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/WsEndpointFilteredResource.java:169]: Pretty unlikely, but I've added an `orElseThrow`.

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+
+    @Inject
+    EmbeddedServer embeddedServer;
+
+    @Test
+    void testWithDownloadOutsideStream() throws Exception {
+
+        var port = embeddedServer.getPort();
+
+        var properties = Map.<String, String>of(
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java:60]: Java compiler is getting good. Can just do `Map.of()` and it gets `<String,String>`. Done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+     * Flowable. But, there is a problem with Java 11:
+     * https://bugs.openjdk.java.net/browse/JDK-8228970 You can view the test here :
+     * https://hg.openjdk.java.net/jdk/jdk/rev/e4cc5231ce2d
+     */
+    @Test
+    @Disabled
+    @SuppressWarnings("java:S1607")
+    void testWithDownloadInsideStream() throws Exception {
+
+        var properties = Map.<String, String>of(
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/FullTest.java:104]: done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/ConfigurationTest.java [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Test
+    void testLoadFilenameNotFound() {
+        assertThatExceptionOfType(ClientException.class).isThrownBy(() -> Configuration.from("doesnotexist"));
+    }
+
+    @Test
+    void testLoadFilenameNull() {
+        assertThatExceptionOfType(IllegalArgumentException.class).isThrownBy(() -> Configuration.from((String) null));
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/ConfigurationTest.java:96]: Did need that as the compiler saw some methods as ambiguous. They are now renamed. I've removed the casts.

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/ConfigurationTest.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+            () -> Configuration.fromDefaults(Map.of("service.url", "\\")));
+        assertThatExceptionOfType(ClientException.class).isThrownBy(
+            () -> Configuration.from(FILENAME, Map.of("service.url", "\\")));
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @Test
+    void testMergeNullSame() {
+        var expected = defaultConfig();
+        var actual = expected.merge((Map) null);
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/abc/impl/ConfigurationTest.java:111]: done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/Configuration.java [hunk 23]
```diff
@@ -0,0 +1,411 @@
...
+            properties.put(KEY_SERVICE_PS_URL, createrPalisadeUrl(baseUri, properties));
+            properties.put(KEY_SERVICE_FRS_URL, createFilteredResourceUrl(baseUri, properties));
+
+        }
+
+        return properties;
+
+    }
+
+    private static String createrPalisadeUrl(final URI baseUri, final Map<String, Object> properties) {
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/Configuration.java:366]: You have sharp eyes!

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/Configuration.java [hunk 24]
```diff
@@ -0,0 +1,411 @@
...
+
+        final var bldr = new StringBuilder()
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/Configuration.java:368]: Done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/dft/DefaultClient.java [hunk 25]
```diff
@@ -0,0 +1,66 @@
...
+
+import uk.gov.gchq.palisade.client.Client;
+import uk.gov.gchq.palisade.client.internal.impl.Configuration;
+import uk.gov.gchq.palisade.client.util.Checks;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * This client is the default impleentation and responds to the subname of
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/dft/DefaultClient.java:26]: Done

File: client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/internal/download/DownloaderTest.java [hunk 26]
```diff
@@ -0,0 +1,100 @@
...
+
+import javax.inject.Inject;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+
+@MicronautTest
+class DownloaderTest {
+
+    private static final String BASE_URL = "http://localhost:"; // needs port added befoire use
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/unit-tests/java/uk/gov/gchq/palisade/client/internal/download/DownloaderTest.java:36]: done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/Query.java [hunk 27]
```diff
@@ -0,0 +1,37 @@
...
+ *
+ * @since 0.5.0
+ */
+public interface Query {
+
+    /**
+     * Executes the query and immediately returns a future which contains the object
+     * which will provide access to returned resources
+     *
+     * @return a {@code CompleteableFuture} containing the result of the query
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/Query.java:32]: done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/Download.java [hunk 28]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * @since 0.5.0
+ */
+public interface Download {
+
+    /**
+     * Returns an {@code InputStream} which contains the stream of bytes for the
+     * file to be downloaded
+     *
+     * @return the inputstream of bytes
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/Download.java:33]: done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/dft/DefaultQueryResponse.java [hunk 29]
```diff
@@ -0,0 +1,221 @@
...
+        this.session = session;
+        this.palisadeResponse = palisadeResponse;
+    }
+
+    @Override
+    public Publisher<Message> stream() {
+
+        // our flowable must wrap the websocket client
+
+        var flowable = Flowable.<Message>create((final FlowableEmitter<Message> emitter) -> {
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/dft/DefaultQueryResponse.java:72]: removed chevrons completely. done

File: client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/ConfigurationException.java [hunk 30]
```diff
@@ -0,0 +1,49 @@
...
+package uk.gov.gchq.palisade.client.internal.impl;
+
+import uk.gov.gchq.palisade.client.ClientException;
+
+/**
+ * An exception which is thrown when a configuration error occurs
+ *
+ * @since 0.5.0
+ */
+public class ConfigurationException extends ClientException {
```
[the remark this answers is not in the document]
Comment (dbz120896) [reply client-java/src/main/java/uk/gov/gchq/palisade/client/internal/impl/ConfigurationException.java:25]: `Configuration` now throws this exception which is a subclass of `ClientException`.

**Pull request closed without merging**